### PR TITLE
Add OpenUSD

### PR DESCRIPTION
**Pull request by Meakk**

### CODE REVIEW

File: .github/actions/f3d-superbuild/action.yml [hunk 1]
```diff
@@ -42,18 +42,20 @@ runs:
     with:
       path: ${{env.SCCACHE_CACHE}}
       key: sccache-cache-${{inputs.raytracing_label}}-${{runner.os}}-0-${{env.DATE_STRING}}
-      restore-keys: sccache-cache-${{inputs.raytracing_label}}-${{runner.os}}-1
+      restore-keys: sccache-cache-${{inputs.raytracing_label}}-${{runner.os}}-2
```
Comment (mwestphal) [inline .github/actions/f3d-superbuild/action.yml:45]: no need to change the cache number,

Comment (mwestphal) [reply .github/actions/f3d-superbuild/action.yml:45]: and I just realised but this should be a zero, not a 1, to be coherent. It is still working as intended.

File: projects/apple/f3d.bundle.cmake [hunk 2]
```diff
@@ -38,6 +38,14 @@ install(
   FILES       "${superbuild_install_location}/f3d.app/Contents/Info.plist"
   DESTINATION "${f3d_appname}/Contents")
 
+# OpenUSD plugin descriptors
+if (openusd_enabled)
+  install(
+    DIRECTORY   "${superbuild_install_location}/lib/usd"
+    DESTINATION "${f3d_appname}/Contents"
+    USE_SOURCE_PERMISSIONS)
+endif()
```
Comment (mwestphal) [inline projects/apple/f3d.bundle.cmake:47]: why only for apple ?

File: projects/win32/f3d.bundle.cmake [hunk 3]
```diff
@@ -31,6 +31,14 @@ if (ospray_enabled)
   endforeach ()
 endif ()
 
+# OpenUSD plugin descriptors
+if (openusd_enabled)
+  install(
+    DIRECTORY   "${superbuild_install_location}/lib/usd"
+    DESTINATION "lib"
+    USE_SOURCE_PERMISSIONS)
+endif()
```
Comment (mwestphal) [inline projects/win32/f3d.bundle.cmake:40]: if this is the same logic with different destination, maybe put in common ? See how its done for the licenses.

**Comment by mwestphal** [comment]
```
2023-11-23T18:36:27.0627698Z FAILED: pxr/base/tf/CMakeFiles/tf.dir/pxrDoubleConversion/bignum-dtoa.cc.obj 
2023-11-23T18:36:27.0643538Z sccache C:\PROGRA~1\MICROS~2\2022\ENTERP~1\VC\Tools\MSVC\1429~1.301\bin\HostX64\x64\cl.exe  /nologo /TP -DBOOST_ALL_DYN_LINK -DBOOST_ALL_NO_LIB -DBOOST_PYTHON_NO_PY_SIGNATURES -DGLX_GLXEXT_PROTOTYPES -DGL_GLEXT_PROTOTYPES -DMFB_ALT_PACKAGE_NAME=tf -DMFB_PACKAGE_MODULE=Tf -DMFB_PACKAGE_NAME=tf -DNOMINMAX -DOPENEXR_DLL -DPXR_BUILD_LOCATION=usd -DPXR_PLUGIN_BUILD_LOCATION=../plugin/usd -DTF_EXPORTS=1 -DWIN32_LEAN_AND_MEAN -DYY_NO_UNISTD_H -D_CRT_SECURE_NO_WARNINGS -D_SCL_SECURE_NO_WARNINGS -Dtf_EXPORTS -ID:\a\f3d-superbuild\f3d-superbuild\build\superbuild\openusd\build\include -external:ID:\a\f3d-superbuild\f3d-superbuild\build\install\include\boost-1_83 -external:ID:\a\f3d-superbuild\f3d-superbuild\build\install\include -external:W0 /EHsc /Zc:rvalueCast /Zc:strictStrings /Zc:inline- /W3 /bigobj /Zi /MP /Gm- /wd4244 /wd4305 /wd4267 /wd4506 /wd4091 /wd4273 /wd4180 /wd4334 /DWIN32 /D_WINDOWS /W3 /GR /EHsc /MD /O2 /Ob2 /DNDEBUG -std:c++14 /showIncludes /Fopxr\base\tf\CMakeFiles\tf.dir\pxrDoubleConversion\bignum-dtoa.cc.obj /Fdpxr\base\tf\CMakeFiles\tf.dir\ /FS -c D:\a\f3d-superbuild\f3d-superbuild\build\superbuild\openusd\src\pxr\base\tf\pxrDoubleConversion\bignum-dtoa.cc
2023-11-23T18:36:27.0662290Z D:\a\f3d-superbuild\f3d-superbuild\build\superbuild\openusd\src\pxr\base\tf\pxrDoubleConversion\bignum-dtoa.cc : fatal error C1090: PDB API call failed, error code '3': D:\a\f3d-superbuild\f3d-superbuild\build\superbuild\openusd\build\pxr\base\tf\CMakeFiles\tf.dir\vc140.pdb
2023-11-23T18:36:27.0667604Z [178/1405] Building CXX object pxr\base\tf\CMakeFiles\tf.dir\pxrDoubleConversion\diy-fp.cc.obj
2023-11-23T18:36:27.0670472Z FAILED: pxr/base/tf/CMakeFiles/tf.dir/pxrDoubleConversion/diy-fp.cc.obj 
2023-11-23T18:36:27.0684316Z sccache C:\PROGRA~1\MICROS~2\2022\ENTERP~1\VC\Tools\MSVC\1429~1.301\bin\HostX64\x64\cl.exe  /nologo /TP -DBOOST_ALL_DYN_LINK -DBOOST_ALL_NO_LIB -DBOOST_PYTHON_NO_PY_SIGNATURES -DGLX_GLXEXT_PROTOTYPES -DGL_GLEXT_PROTOTYPES -DMFB_ALT_PACKAGE_NAME=tf -DMFB_PACKAGE_MODULE=Tf -DMFB_PACKAGE_NAME=tf -DNOMINMAX -DOPENEXR_DLL -DPXR_BUILD_LOCATION=usd -DPXR_PLUGIN_BUILD_LOCATION=../plugin/usd -DTF_EXPORTS=1 -DWIN32_LEAN_AND_MEAN -DYY_NO_UNISTD_H -D_CRT_SECURE_NO_WARNINGS -D_SCL_SECURE_NO_WARNINGS -Dtf_EXPORTS -ID:\a\f3d-superbuild\f3d-superbuild\build\superbuild\openusd\build\include -external:ID:\a\f3d-superbuild\f3d-superbuild\build\install\include\boost-1_83 -external:ID:\a\f3d-superbuild\f3d-superbuild\build\install\include -external:W0 /EHsc /Zc:rvalueCast /Zc:strictStrings /Zc:inline- /W3 /bigobj /Zi /MP /Gm- /wd4244 /wd4305 /wd4267 /wd4506 /wd4091 /wd4273 /wd4180 /wd4334 /DWIN32 /D_WINDOWS /W3 /GR /EHsc /MD /O2 /Ob2 /DNDEBUG -std:c++14 /showIncludes /Fopxr\base\tf\CMakeFiles\tf.dir\pxrDoubleConversion\diy-fp.cc.obj /Fdpxr\base\tf\CMakeFiles\tf.dir\ /FS -c D:\a\f3d-superbuild\f3d-superbuild\build\superbuild\openusd\src\pxr\base\tf\pxrDoubleConversion\diy-fp.cc
2023-11-23T18:36:27.0702271Z D:\a\f3d-superbuild\f3d-superbuild\build\superbuild\openusd\src\pxr\base\tf\pxrDoubleConversion\diy-fp.cc: fatal error C1041: cannot open program database 'D:\a\f3d-superbuild\f3d-superbuild\build\superbuild\openusd\build\pxr\base\tf\CMakeFiles\tf.dir\vc140.pdb'; if multiple CL.EXE write to the same .PDB file, please use /FS
```

Not a windows expert but that looks like some Debug/Release issue.

**Comment by mwestphal** [comment]
```
Coding Error: in _GetPluginForType at line 147 of /Users/runner/work/f3d-superbuild/f3d-superbuild/build/superbuild/openusd/src/pxr/usd/ar/resolver.cpp -- Failed to find plugin for ArDefaultResolver
```

Let me know if you want me to try on my mac VM

File: projects/win32/f3d.bundle.cmake (hunk 3 above)
Comment (Meakk) [reply projects/win32/f3d.bundle.cmake:40]: I tried that but for some reason it's failing, I'm not sure why yet.

File: .github/actions/f3d-superbuild/action.yml (hunk 1 above)
Comment (Meakk) [reply .github/actions/f3d-superbuild/action.yml:45]: I changed the MSVC version used, I think it's necessary to invalidate the cache, otherwise I have link errors.

**Comment by Meakk** [comment]
> ```
> Coding Error: in _GetPluginForType at line 147 of /Users/runner/work/f3d-superbuild/f3d-superbuild/build/superbuild/openusd/src/pxr/usd/ar/resolver.cpp -- Failed to find plugin for ArDefaultResolver
> ```
> 
> Let me know if you want me to try on my mac VM

Yes please. Basically I put the `usd` folder randomly in `Contents` but it's probably not the right location. I'd say it should be in either in the application folder, or in the `libxxxUsd.dylib` folder.

File: .github/actions/f3d-superbuild/action.yml (hunk 1 above)
Comment (mwestphal) [reply .github/actions/f3d-superbuild/action.yml:45]: you can now update the superbuild instead

File: .github/actions/f3d-superbuild/action.yml [hunk 4]
```diff
@@ -85,6 +87,7 @@ runs:
       -DBUILD_SHARED_LIBS_openexr=OFF
       -DBUILD_SHARED_LIBS_imath=OFF
       -DBUILD_SHARED_LIBS_occt=OFF
+      -DBUILD_SHARED_LIBS_openusd=ON
```
Comment (mwestphal) [inline .github/actions/f3d-superbuild/action.yml:90]: is that necessary ?

File: projects/apple/f3d.bundle.cmake [hunk 5]
```diff
@@ -38,6 +38,14 @@ install(
...
   DESTINATION "${f3d_appname}/Contents")
 
+# OpenUSD plugin descriptors
+if (openusd_enabled)
+  install(
+    DIRECTORY   "${superbuild_install_location}/lib/usd"
+    DESTINATION "${f3d_appname}/Contents/MacOS"
+    USE_SOURCE_PERMISSIONS)
+endif()
+
```
Comment (mwestphal) [inline projects/apple/f3d.bundle.cmake:48]: Its building in my VM, Ill let you know.

**Comment by mwestphal** [comment]
```

Thread 0 Crashed:: Dispatch queue: com.apple.main-thread
0   libf3d.2.2.dylib              	0x000000011129324d pxrInternal_v0_23__pxrReserved__::(anonymous namespace)::_GetAvailablePrimaryResolvers(std::__1::vector<pxrInternal_v0_23__pxrReserved__::(anonymous namespace)::_ResolverInfo, std::__1::allocator<pxrInternal_v0_23__pxrReserved__::(anonymous namespace)::_ResolverInfo> > const&) + 605
1   libf3d.2.2.dylib              	0x0000000111294e0b pxrInternal_v0_23__pxrReserved__::(anonymous namespace)::_DispatchingResolver::_DispatchingResolver() + 539
2   libf3d.2.2.dylib              	0x0000000111292b83 pxrInternal_v0_23__pxrReserved__::ArGetResolver() + 51
3   libf3d.2.2.dylib              	0x000000011113d229 pxrInternal_v0_23__pxrReserved__::SdfLayer::_ComputeInfoToFindOrOpenLayer(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > > const&, pxrInternal_v0_23__pxrReserved__::SdfLayer::_FindOrOpenLayerInfo*, bool) + 329
4   libf3d.2.2.dylib              	0x000000011113db5a pxrInternal_v0_23__pxrReserved__::SdfLayer::FindOrOpen(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > > const&) + 298
5   libf3d.2.2.dylib              	0x0000000110cfc64e pxrInternal_v0_23__pxrReserved__::_OpenLayer(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, pxrInternal_v0_23__pxrReserved__::ArResolverContext const&) + 238
6   libf3d.2.2.dylib              	0x0000000110cfc21c pxrInternal_v0_23__pxrReserved__::UsdStage::Open(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, pxrInternal_v0_23__pxrReserved__::UsdStage::InitialLoadSet) + 316
7   libf3d.2.2.dylib              	0x000000010f708d79 vtkF3DUSDImporter::vtkInternals::ReadScene(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) + 25
8   libf3d.2.2.dylib              	0x000000010f708d54 vtkF3DUSDImporter::ImportBegin() + 20
```

File: projects/win32/f3d.bundle.cmake (hunk 3 above)
Comment (mwestphal) [reply projects/win32/f3d.bundle.cmake:40]: wants me to take a look ?

File: projects/win32/f3d.bundle.cmake [hunk 6]
```diff
@@ -68,7 +76,7 @@ if (cpack_generator MATCHES "NSIS")
   file(COPY
       "${superbuild_source_directory}/resources/logo.ico"
       "${superbuild_source_directory}/resources/logotype64.bmp"
-    DESTINATION "${superbuild_install_location}/share/f3d/")
+    DESTINATION "${superbuild_install_location}/share/f3d")
```
Comment (mwestphal) [inline projects/win32/f3d.bundle.cmake:79]: looks unrelated, is t was not working before ?

File: projects/win32/f3d.bundle.cmake (hunk 3 above)
Comment (mwestphal) [reply projects/win32/f3d.bundle.cmake:40]: ping

**Comment by mwestphal** [comment]
Can I do anything to help move this forward ?

File: projects/win32/f3d.bundle.cmake (hunk 3 above)
Comment (Meakk) [reply projects/win32/f3d.bundle.cmake:40]: Let me fix the remaining issues and we'll see that after.

File: projects/f3d.bundle.common.cmake [hunk 7]
```diff
@@ -56,3 +56,17 @@ function (f3d_package_all_licenses)
...
+
+  if (NOT DEFINED USD_PLUGINS_DESTINATION)
+    message(FATAL_ERROR "f3d_package_usd_plugins is called without DESTINATION")
+  endif ()
+
+  install(
+    DIRECTORY   "${superbuild_install_location}/lib/usd"
+    DESTINATION "${USD_PLUGINS_DESTINATION}"
+    USE_SOURCE_PERMISSIONS)
+endfunction ()
```
Comment (mwestphal) [inline projects/f3d.bundle.common.cmake:72]: missing eol

File: projects/openusd.cmake [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+    -DPXR_BUILD_TUTORIALS:BOOL=OFF
+    -DPXR_BUILD_USD_TOOLS:BOOL=OFF
+    -DPXR_ENABLE_GL_SUPPORT:BOOL=OFF
+    -DPXR_ENABLE_PYTHON_SUPPORT:BOOL=OFF
+    -DPXR_INSTALL_LOCATION:STRING=../lib/usd
+)
+
+superbuild_apply_patch(openusd onetbb "Add support for oneTBB")
+superbuild_apply_patch(openusd msvc-defaults "Remove PDB generation and parallel build")
+superbuild_apply_patch(openusd install-bin "Install Windows DLL in bin directory")
```
Comment (mwestphal) [inline projects/openusd.cmake:20]: it would be nice to add links to github issues for each patch

File: projects/win32/f3d.bundle.cmake [hunk 9]
```diff
@@ -5,7 +5,7 @@ include(f3d.bundle.common)
 ## Package binaries
 
 # Set where library will be found by fixup_bundle.py
-set(library_paths "${superbuild_install_location}/bin")
+set(library_paths "${superbuild_install_location}/bin" "${superbuild_install_location}/lib")
```
Comment (mwestphal) [inline projects/win32/f3d.bundle.cmake:8]: didnt you patch that out ?

**Review by mwestphal — CHANGES_REQUESTED** [review]
small changes needed

File: projects/win32/f3d.bundle.cmake (hunk 9 above)
Comment (Meakk) [reply projects/win32/f3d.bundle.cmake:8]: yeah good catch

**Review by mwestphal — APPROVED** [review]

**Review by jeffersonrichard135-afk — APPROVED** [review]

File: .github/actions/wheels/action.yml
Comment (jeffersonrichard135-afk) [inline .github/actions/wheels/action.yml:1]: Yo

**Review by jeffersonrichard135-afk — CHANGES_REQUESTED** [review]

File: projects/openusd.cmake (hunk 8 above)
Comment (jeffersonrichard135-afk) [reply projects/openusd.cmake:25]: > it would be nice to add links to github issues for each patch

File: .github/actions/wheels/action.yml
Comment (mwestphal) [reply .github/actions/wheels/action.yml:1]: Whats up @jeffersonrichard135-afk ?

Comment (jeffersonrichard135-afk) [reply .github/actions/wheels/action.yml]: Good how can I do it
